src/main/java/org/datadog/jenkins/plugins/datadog/clients/HttpRetryPolicy.java

```diff
@@ -0,0 +1,130 @@
+ *             for 5XX responses)
+ *       </ul>
+ *   <li>in all other cases do not retry
+ * </ul>
+ *
+ * <p>Instances of this class are not thread-safe and not reusable: each HTTP call requires its own
+ * instance.
+ */
+@NotThreadSafe
+public class HttpRetryPolicy {
+
+  private static final Logger logger = Logger.getLogger(HttpRetryPolicy.class.getName());
+
+  private static final int NO_RESPONSE_RECEIVED = -1;
+  private static final int TOO_MANY_REQUESTS_HTTP_CODE = 429;
+  private static final String X_RATELIMIT_RESET_HTTP_HEADER = "x-ratelimit-reset";
+  private static final int RATE_LIMIT_RESET_TIME_UNDEFINED = -1;
+  private static final int MAX_ALLOWED_WAIT_TIME_SECONDS = 10;
+  private static final int RATE_LIMIT_DELAY_RANDOM_COMPONENT_MAX_MILLIS = 401;
+
+  private int retriesLeft;
+  private long delay;
+  private final double delayFactor;
+
+  private HttpRetryPolicy(int retriesLeft, long delay, double delayFactor) {
+    this.retriesLeft = retriesLeft;
+    this.delay = delay;
+    this.delayFactor = delayFactor;
+  }
+
+  public boolean shouldRetry(@Nullable Response response) {
+    if (retriesLeft == 0) {
+      return false;
+    }
+
+    int responseCode = response != null ? response.getStatus() : NO_RESPONSE_RECEIVED;
+    if (responseCode == TOO_MANY_REQUESTS_HTTP_CODE) {
+      long waitTimeSeconds = getRateLimitResetTime(response);
+      if (waitTimeSeconds == RATE_LIMIT_RESET_TIME_UNDEFINED) {
+        retriesLeft--; // doing a regular retry if proper reset time was not provided
+        return true;
+      }
+
+      if (waitTimeSeconds > MAX_ALLOWED_WAIT_TIME_SECONDS) {
+        return false; // too long to wait, will not retry
+      }
+
+      retriesLeft = 0;
+      delay =
+          TimeUnit.SECONDS.toMillis(waitTimeSeconds)
+              + ThreadLocalRandom.current().nextInt(RATE_LIMIT_DELAY_RANDOM_COMPONENT_MAX_MILLIS);
+      return true;
+
+    } else if (responseCode >= 500 || responseCode == NO_RESPONSE_RECEIVED) {
+      retriesLeft--;
+      return true;
+
+    } else {
+      return false;
+    }
+  }
+
+  private long getRateLimitResetTime(Response response) {
+    String rateLimitHeader = response.getHeaders().get(X_RATELIMIT_RESET_HTTP_HEADER);
+    if (rateLimitHeader == null) {
+      return RATE_LIMIT_RESET_TIME_UNDEFINED;
+    }
+
+    try {
+      return Long.parseLong(rateLimitHeader);
+    } catch (NumberFormatException e) {
+      DatadogUtilities.severe(logger, e,
+              "Could not parse "
+              + X_RATELIMIT_RESET_HTTP_HEADER
+              + " header contents: "
+              + rateLimitHeader);
+      return RATE_LIMIT_RESET_TIME_UNDEFINED;
+    }
+  }
+
+  public long backoff() {
+    long currentDelay = delay;
+    delay = (long) (delay * delayFactor * ThreadLocalRandom.current().nextDouble(0.85, 1.15));
+    return currentDelay;
+  }
+
+  public static final class Factory {
+    private final int maxRetries;
+    private final long initialDelay;
```

Comment:
Maybe just `delay`? It's confusing to find `initial` as it seems it's only going to be used once, but this is the actual delay for every backoff delay.

Reply:
The field in the policy class is mutable and is called `delay`, but the one in the factory class is really just the initial value